_posts/2023-01-23-there-is-no-cloud.md

```diff
@@ -0,0 +1,294 @@
+on the AWS DynamoDB side:
+
+```
+{
+  "0-pk": {
+    "S": "partition-value-01"
+  },
+  "1-sk": {
+    "S": "1"
+  },
+  "*amzn-ddb-map-desc*": {
+    "B": "AAAAAAAAABBhbXpuLWRkYi1lbnYtYWxnAAAAB0FFUy8yNTYAAAAQYW16bi1kZGItZW52LWtleQAAADhaNUswVGxMV0JtS2VreDZwc0lOYlB0THphV0pxOGFlLytKNk9JMFVQVnhMNXk5Q1BLdG5RT1E9PQAAABdhbXpuLWRkYi1tYXAtc2lnbmluZ0FsZwAAAApIbWFjU0hBNTEyAAAAFWFtem4tZGRiLW1hcC1zeW0tbW9kZQAAABEvQ0JDL1BLQ1M1UGFkZGluZwAAABFhbXpuLWRkYi13cmFwLWFsZwAAAAdBRVNXcmFw"
+  },
+  "*amzn-ddb-map-sig*": {
+    "B": "3AZffQmyNl0w7QcznfBvCy2TfRmyNfOqnITu1hrO/JDc2M1hAL01E0yqIJSgFtkISIk5TtXdaZoXAfOjN6jHiw=="
+  },
+  "example": {
+    "B": "ITFfW5lICGQIoK29WvZoh0cmwaKteDack3KmgXBZxCo="
+  },
+  "numbers": {
+    "B": "kgPie2XmokAps6XDKx3KwHQEkq0AtlfhoqT8joSQ6xE="
+  },
+  "test": {
+    "S": "test-value"
+  }
+}
+```
+First off, you see the partition `0-pk` and sort `1-sk` keys UNENCRYPTED.
+Second, remember that the `test` attribute was UNENCRYPTED for fun.
+Third, most importantly, all other attributes, namely, `example` and `numbres` 
+are duly ENCRYPTED! We also see a few of service `*amzn` fields, which
+describe encryption.
+
+All right, I revealed everything except how I created my DynamoDB table:
+
+```
+table = client.create_table(
+        TableName=table_name,
+        KeySchema=[
+            {
+                'AttributeName': '0-pk',
+                'KeyType': 'HASH'  # Partition key
+            },
+            {
+                'AttributeName': '1-sk',
+                'KeyType': 'RANGE'  # Sort key
+            }
+        ],
+        AttributeDefinitions=[
+            {
+                'AttributeName': '0-pk',
+                'AttributeType': 'S'
+            },
+            {
+                'AttributeName': '1-sk',
+                'AttributeType': 'S'
+            }
+        ],
+        ProvisionedThroughput={
+            'ReadCapacityUnits': 10,
+            'WriteCapacityUnits': 10
+        }
+    )
+```
+Here I just define my primary (partition/hash + sort/range) keys, 
+NoSQL business as usual.
+
+And the last most important thing - here is how I generate my wrapping and 
+signing keys, ON PREMISES, and never give them away:
+
+```
+def create_asymmetric_keys():
+    wrapping_key_bytes = JceNameLocalDelegatedKey.generate("RSA", 4096).key
+    signing_key_bytes = JceNameLocalDelegatedKey.generate("SHA512withRSA", 4096).key
+    file_name = 'key_asymm_wrap.bin'
+    with open(file_name, 'wb') as f:
+        f.write(wrapping_key_bytes)
+    os.chmod(file_name, 0o600)
+    file_name = 'key_asymm_sign.bin'
+    with open(file_name, 'wb') as f:
+        f.write(signing_key_bytes)
+    os.chmod(file_name, 0o600)
+```
+This is as simple as possible, for PoC, because in production you need to 
+organize for a proper PKI. Similarly, you may use symmetric keys instead
+of asymmetric ones if you want.
+
+
+## That's All, Folks!
+
+## Or Not?
+
+## Your Next Friend is FHE - Fully Homomorphic Encryption
```

Comment:
formatting, three consecutive headers without any text.

Reply:
These are NOT sections, just bold font, bigger size

Reply:
in markdown `##` is a level 2 heading

Reply:
but makes bigger nice bold as well, bigger that **bold**